### PR TITLE
Add unconstrained benchmarks.

### PR DESCRIPTION
**Pull request by pkofod**
More or less copied from https://github.com/JuliaOpt/Optim.jl/pull/259

Certainly this can, and should, be simplified. It should also be extended to constrained optimization and show relevant convergence info in that case. Currently produces the following plots
**CUTEst test bed**
f_err_cutest
x_err_cutest

**Optim test bed**
f_err_optim
x_err_optim

**TODO**
- [ ] Register "minimum" if available in CUTEst.
- [ ] Figure out the best way to log run time; `@benchmarks`?
- [ ] (maybe not this time around) find an easy way to compare two commits or different solvers
- [ ] Add tests for different linesearches.

### CODE REVIEW

**Comment by pkofod** [comment]
Obviously doesnt work without the constrained branch

File: benchmarks/unconstrained/run_benchmarks.jl [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    run(`mkdir $version_dir`)
+catch
+
+end
+cd(version_dir)
+
+default_names = ["Accelerated Gradient Descent",
+                 "BFGS",
+                 "L-BFGS",
+                 "Momentum Gradient Descent",
```
Comment (anriseth) [inline benchmarks/unconstrained/run_benchmarks.jl:20]: This position in default_solvers points to `ConjugateGradient()`, not to `MomentumGradientDescent()`

File: benchmarks/unconstrained/default/cutest_benchmarks.jl [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    nlp = CUTEstModel(prob)
+    d, constraints = optim_problem(nlp)
+    x0 = initial_x(nlp)
+
+    x_hat = copy(x0)
+    f_hat = obj(nlp, x_hat)
+    xs = []
+    times = []
+        for i = 1:n
+            if !(default_solvers[i] in (Newton(), NewtonTrustRegion()))
```
Comment (anriseth) [inline benchmarks/unconstrained/default/cutest_benchmarks.jl:32]: Why do you not run these tests with the Newton solvers?

It worked well for me to remove this if statement, and use the following code to get results and mintime:

```jl
result = optimize(d, x0, default_solvers[i], options)
mintime = @elapsed optimize(d, x0, default_solvers[i], options)
```

File: benchmarks/unconstrained/default/cutest_benchmarks.jl [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                                   Inf,
+                                   Inf,
+                                   Inf,
+                                   Inf]))
+                    push!(xs, fill(Inf, length(x0)))
+            	end
+            end
+        end
+        for i = 1:n
+            if !(default_solvers[i] in (Newton(), NewtonTrustRegion())) # should be typeof-ish instead
```
Comment (anriseth) [inline benchmarks/unconstrained/default/cutest_benchmarks.jl:63]: Why not include Newton? Should work to just remove this statement with the code above.

File: benchmarks/unconstrained/run_benchmarks.jl (hunk 1 above)
Comment (pkofod) [reply benchmarks/unconstrained/run_benchmarks.jl:20]: Might be safer to pass in tuples of (name, solver) for this very reason!

File: benchmarks/unconstrained/default/cutest_benchmarks.jl (hunk 2 above)
Comment (pkofod) [reply benchmarks/unconstrained/default/cutest_benchmarks.jl:32]: I think I might have used problems in `C^1` at some point, but as you can see it is now `C^n` for n>=2, so should definitely add them back in.

File: benchmarks/unconstrained/default/cutest_benchmarks.jl (hunk 3 above)
Comment (pkofod) [reply benchmarks/unconstrained/default/cutest_benchmarks.jl:63]: As mentioned above, I cannot really remember why I did it. It was largely copy+paste from Optim, and I can't remember why I did it there. Thanks for pointing it out!

**Pull request closed without merging**